runtime/cudaq/algorithms/evolve.h

```diff
@@ -11,145 +11,242 @@
-      auto prepare_state = [intermediate_states]() {
-        auto qs = qvector<2>(intermediate_states.back());
-      };
-      for (auto observable : observables[++step_idx]) {
-        shots_count <= 0
-            ? expectations.push_back(observe(prepare_state, observable))
-            : expectations.push_back(
-                  observe(shots_count, prepare_state, observable));
-      }
-      expectation_values.push_back(expectations);
-    }
-  }
-  if (step_idx < 0)
-    return evolve_result(intermediate_states);
-  return evolve_result(intermediate_states, expectation_values);
+template <typename OpTy>
+std::vector<cudaq::operator_sum<cudaq::matrix_operator>>
+convertOps(const std::vector<OpTy> &ops) {
+  std::vector<cudaq::operator_sum<cudaq::matrix_operator>> converted;
+  for (const auto &op : ops)
+    converted.emplace_back(convertOp(op));
+  return converted;
 }
 
-template <typename QuantumKernel>
-async_evolve_result
-evolve_async(state initial_state, QuantumKernel &&kernel,
-             const std::vector<spin_op> &observables = {},
-             std::size_t qpu_id = 0,
-             std::optional<cudaq::noise_model> noise_model = std::nullopt,
-             int shots_count = -1) {
-  auto &platform = cudaq::get_platform();
-  std::promise<evolve_result> promise;
-  auto f = promise.get_future();
-
-  QuantumTask wrapped = detail::make_copyable_function(
-      [p = std::move(promise), func = std::forward<QuantumKernel>(kernel),
-       initial_state, observables, noise_model, shots_count,
-       &platform]() mutable {
-        if (noise_model.has_value())
-          platform.set_noise(&noise_model.value());
-        p.set_value(evolve(initial_state, func, observables, shots_count));
-        if (noise_model.has_value())
-          platform.set_noise(nullptr);
-      });
-
-  platform.enqueueAsyncTask(qpu_id, wrapped);
-  return f;
+template <typename OpTy>
+std::vector<cudaq::operator_sum<cudaq::matrix_operator>>
+convertOps(const std::initializer_list<OpTy> &ops) {
+  std::vector<cudaq::operator_sum<cudaq::matrix_operator>> converted;
+  for (const auto &op : ops)
+    converted.emplace_back(convertOp(op));
+  return converted;
 }
+} // namespace __internal__
 
-template <typename QuantumKernel>
-async_evolve_result
-evolve_async(state initial_state, std::vector<QuantumKernel> kernels,
-             const std::vector<std::vector<spin_op>> &observables = {},
-             std::size_t qpu_id = 0,
-             std::optional<cudaq::noise_model> noise_model = std::nullopt,
-             int shots_count = -1) {
-  auto &platform = cudaq::get_platform();
-  std::promise<evolve_result> promise;
-  auto f = promise.get_future();
+template <typename HamTy,
+          typename CollapseOpTy = cudaq::operator_sum<cudaq::matrix_operator>,
+          typename ObserveOpTy = cudaq::operator_sum<cudaq::matrix_operator>,
+          typename = std::enable_if_t<cudaq::operator_type<HamTy> &&
```

Comment:
might be reading this wrong - but make sure you `#ifdef` here on the C++ standard. You want this new enable_if_t if it is C++17, and the requires clause if it is C++20, but you do not want both